json_annotation/lib/src/checked_helpers.dart

```diff
@@ -13,11 +13,22 @@ typedef _CastFunction<R> = R Function(Object?);
 T $checkedCreate<T>(
   String className,
   Map map,
-  T Function(S Function<S>(String, _CastFunction<S>) converter) constructor, {
+  T Function(
+    S Function<S>(
+      String,
+      _CastFunction<S>, {
+      Object? Function(Map, String)? readValue,
```

Comment:
Should `readValue` also return a `T` here to give better static checking? In general the relationship here between the `_CastFunction` and `readValue` params is a bit murky to me.

Reply:
Will investigate...

Reply:
So this is a bit weird.

`readValue` is meant to be pre-converter. So it should return a JSON-ish result. that then goes through any associated converters, etc.

If one wants `readValue` to return the field type, they have to also create a converter.

I pondered doing the logic where `readValue` just takes over ALL conversion logic, too. But that gets tricky.

Reply:
Ya, that is what I was wondering about. This would cover some number of use cases, where one or more json fields can be merged into a new json-like field, but its a bit indirect for others. For instance you might have people actually returning a Map from `readValue`, built from some other fields, and then deserializing that map into some class in the regular conversion function.

Ultimately I think this choice is probably fine, but I would clarify things in the documentation?

Reply:
Added more docs. PTAL

Reply:
> I pondered doing the logic where `readValue` just takes over ALL conversion logic, too. But that gets tricky.

Tricky for users or for this library?

It feels like a simpler concept to document if `readValue` does all the work instead of coordinating with some other uncoupled concept with no static types in between...

Reply:
It could be both. Imagine a case where I just want to rename a field `foo` -> `bar` but the type is something complex like `Map<int, List<Duration?>>?` – that's a bear to rewrite for the user.